Add utilities to convert from html to htpy

OleJoik opened the pull request:
# Convert HTML to htpy code

Maybe you already have a bunch of HTML, or templates that you would like to migrate to htpy. 
We got you covered. The utility command `html2htpy` ships with `htpy`, and can be used to transform existing 
html into Python code (htpy!).

```
$ html2htpy -h
usage: html2htpy [-h] [-f {auto,ruff,black,none}] [-i {yes,h,no}] [--no-shorthand] [input]

positional arguments:
  input                 input HTML from file or stdin

options:
  -h, --help            show this help message and exit
  -f {auto,ruff,black,none}, --format {auto,ruff,black,none}
                        Select one of the following formatting options: auto, ruff, black or none
  -i {yes,h,no}, --imports {yes,h,no}
                        Output mode for imports of found htpy elements
  --no-shorthand        Use explicit `id` and `class_` kwargs instead of the shorthand #id.class syntax
```


Lets say you have an existing HTML file:

```html
<!-- index.html -->

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>htpy Recipes</title>
</head>
<body>
    <div id="header">
        <h1>Welcome to the cooking site</h1>
        <p>Your go-to place for delicious recipes!</p>
    </div>

    <div id="recipe-of-the-day" class="section">
        <h2>Recipe of the Day: <span class="highlight">Spaghetti Carbonara</span></h2>
        <p>This classic Italian dish is quick and easy to make.</p>
    </div>

    <div id="footer">
        <p>&copy; 2024 My Cooking Site. All rights reserved.</p>
    </div>
</body>
</html>
```

Now, if you run the command, it outputs the corresponding Python code (htpy).

```
$  html2htpy index.html
```

```py
from htpy import body, div, h1, h2, head, html, meta, p, span, title

html(lang="en")[
    head[
        meta(charset="UTF-8"),
        meta(name="viewport", content="width=device-width, initial-scale=1.0"),
        title["htpy Recipes"],
    ],
    body[
        div("#header")[
            h1["Welcome to the cooking site"], p["Your go-to place for delicious recipes!"]
        ],
        div("#recipe-of-the-day.section")[
            h2["Recipe of the Day: ", span(".highlight")["Spaghetti Carbonara"]],
            p["This classic Italian dish is quick and easy to make."],
        ],
        div("#footer")[p["© 2024 My Cooking Site. All rights reserved."]],
    ],
]
```

## Piping input/stdin stream

You can also pipe input to htpy, for example `cat demo.html | html2htpy`.

This can be combined with other workflows in the way that you find most suitable. 
For example, you might pipe from your clipboard to htpy, and optionally direct the output to a file.

#### Linux 

```
xclip -o -selection clipboard | html2htpy > output.py
```

#### Mac 

```
pbpaste | html2htpy > output.py
```

#### Windows

```
powershell Get-Clipboard | html2htpy > output.py
```


## Formatting the output
`html2htpy` can format the output Python code using `black` or `ruff`.
Select the preferred formatter with the `-f`/`--format` flag. Options are `auto`, `ruff`, `black` and `none`.

By default, the selection will be `auto`, formatting if it finds a formatter on path, preffering `ruff` if it's available.
If no formatters are available on path, the output will not be formatted.


## Import options

You have a couple of options regarding imports with the `-i`/`--imports` flag. 
Options are `yes` (default), `h`, `no`. 

#### Module import of htpy: `--imports=h`

Some people prefer to `import htpy as h` instead of importing individual elements from htpy.
If this is you, you can use the `--imports=h` option to get corresponding output when using `html2htpy`.

```
$ html2htpy --imports=h example.html
```
```py
import htpy as h

h.section("#main-section.hero.is-link")[
    h.p(".subtitle.is-3.is-spaced")["Welcome"]
]
```

## Explicit id and class kwargs


If you prefer the explicit `id="id", class_="class"` kwargs syntax over the default htpy shorthand `#id.class` syntax, you can get it by passing the `--no-shorthand` flag.

```html
<!-- example.html -->
<section id="main-section" class="hero is-link">
    <p class="subtitle is-3 is-spaced">Welcome</p>
</section>
```

#### Default shorthand yield `#id.class`

```
$ html2htpy example.html
```
```py
from htpy import p, section

section("#main-section.hero.is-link")[
    p(".subtitle.is-3.is-spaced")["Welcome"]
]
```

#### No shorthand yields kwargs `id`, `class_`
```
$ html2htpy --no-shorthand example.html
```
```py
from htpy import p, section

section(id="main-section", class_="hero is-link")[
    p(class_="subtitle is-3 is-spaced")["Welcome"]
]
```


## Template interpolation to f-strings

`html2htpy` will try to convert template variables to pythonic f-strings:

`template {{ variables }}` -> `f"template { variables }"`

Note that other typical template syntax, such as loops `{% for x in y %}`, can not be transformed this way, 
so you will often have to clean up a bit after `html2htpy` is done with its thing.

See the example below:

```html
<!-- jinja.html -->
<body>
    <h1>{{ heading }}</h1>
    <p>Welcome to our cooking site, {{ user.name }}!</p>

    <h2>Recipe of the Day: {{ recipe.name }}</h2>
    <p>{{ recipe.description }}</p>

    <h3>Instructions:</h3>
    <ol>
        {% for step in recipe.steps %}
        <li>{{ step }}</li>
        {% endfor %}
    </ol>
</body>
```
```
$ html2htpy jinja.html
```
```py
from htpy import body, h1, h2, h3, li, ol, p

body[
    h1[f"{ heading }"],
    p[f"Welcome to our cooking site, { user.name }!"],
    h2[f"Recipe of the Day: { recipe.name }"],
    p[f"{ recipe.description }"],
    h3["Instructions:"],
    ol[
        """        {% for step in recipe.steps %}        """,
        li[f"{ step }"],
        """        {% endfor %}    """,
    ],
]
```

OleJoik commented:
Might be a better idea to have it parse files than to paste into the terminal, bit of jank but atleast it works.

pelme reviewed (changes requested):
This is GREAT! Thanks a lot for this nice command! Really looking forward to be using this.

OleJoik commented on htpy/cli/cli.py:
```diff
@@ -0,0 +1,79 @@
+
+    convert_parser = subparsers.add_parser(
+        "convert", help="convert html to python (htpy)"
+    )
```
Absolutely, then I put both the cli and the python code into `htpy/html2htpy.py` and renamed both the command and the function to html2htpy.

OleJoik replied:
In the current state, I import `html2htpy` directly from `htpy` as such: `from htpy import html2htpy`. But there is a slight issue. My tooling thinks that the type is not function (as it should be), but instead thinks it is an `Element`, due to your neat module level `__getattr__` trick. 

I actually getting the function, not a htpy `Element`, it is just the typing that `lies`. Yet it is probably going to be confusing for users. Maybe it is still better to keep a `utils` module to separate "other" imports, such as this one `from htpy.utils import html2htpy` ... from Elements imported as is from `htpy`? Another alternative might be to import elements from a module `from htpy.elements import div, p` but this is obviously breaking and painful...  

![image](https://github.com/pelme/htpy/assets/57186239/3f4565ca-7da7-418c-be59-45b2f0ddd4dd)

OleJoik replied:
Another alternative might be to just import the util function as such: `from htpy.html2htpy import html2htpy`

OleJoik replied:
I made some comments here previously that can safely be disregarded...

OleJoik commented on htpy/utils/html_to_htpy.py:
```diff
@@ -0,0 +1,208 @@
+        _attrs = ""
+        _kwattrs: list[tuple[str, str | None]] = []
+
+        for i in range(len(self.attrs)):
```
Yeah, I updated this. Was a remnant after modifying the algorithm.

OleJoik commented on tests/test_convert_html.py:
```diff
@@ -0,0 +1,271 @@
+            div(class_="hero-body")[
+                div(class_="container")[
+                    p(class_="subtitle is-3 is-spaced")["Welcome"],
+                    p(class_="title is-1 is-spaced")[f"Student code: {student_code}"]
```
While flashy, it comes with a couple of challanges, I realize after the fact: 
1. a need to escape certain symbols that may exist in the content, e.g `"`, `{` and `}`. There might also be more I'm forgetting (?). Ill add these in a commit shortly.
2. Currently this is only done to content, not for example to attributes. Do you think that is sufficient, or would you prefer doing this to props as well?

OleJoik replied:
Maybe it would be better to try to use `'` or `"` (whichever is not used in the content) if one is 'available', and only escape in the last case, where both of them are already in use. But I don't know, don't feel like I personally use double quotes in html content that often so maybe its not terrible that they are escaped by the tool. Regarding the `"` symbol this obviously also applies to all content, not just template expressions.

OleJoik replied:
I believe it's covered now by this test:

```python
def test_convert_f_string_escaping():
    input = """
        <p>{{ variable }} is "a" { paragraph }.</p>
    """

    actual = html2htpy(input, format=False)
    expected = r'p[f"{ variable } is \"a\" {{ paragraph }}."]'

    assert actual == expected
```

pelme replied:
I think it is fine even though it may not be perfect. We can always improve it in the future too. It is not crucial to get all edge cases right, the generated code would still need to be adopted by hand to be fully usable.

OleJoik commented on tests/test_convert_html.py:
```diff
@@ -0,0 +1,271 @@
+    assert expected_output == actual_output
+
+
+def _format(s: str):
```
I had some problems with the indentation of the multiline strings, thats why I added them. However, found another solution that I think is better, using `textwrap.dedent` to align the `expected` multiline strings with the first line. I agree that the formatting there made it all a bit fuzzy.

pelme replied:
nice, `textwrap.dedent` sounds like a good solution for this!

pelme commented on htpy/__init__.py:
```diff
@@ -10,6 +10,8 @@
 from markupsafe import escape as _escape
 
+from .html2htpy import *
+
```
Is this import needed? I think we can just remove it? And then import like `from htpy.html2htpy import html2htpy`? I think mypy/type checkers should pick up that properly! I think very few people will import the html2htpy function, this feature will mostly be used from the cli anyways. But it is nice to have a python API for it too.

OleJoik replied:
In this case, it is indeed not needed! I will remove it.

pelme reviewed (changes requested):
I had another look and left some comments. I think we should have a separate page in the docs with some examples too :)

OleJoik commented on htpy/utils/html_to_htpy.py:
```diff
@@ -0,0 +1,208 @@
+                o, mode=black.FileMode(line_length=80, magic_trailing_comma=False)
+            )
+        else:
+            return o
```
I was looking into this a bit, and decided to drop the formatting stuff for now. I do find it interesting though!

A couple of notes. Disclaimer: I have not personally started using ruff yet, although I probably should. So I don't know ins and outs.

1. I did not find a python interface for ruff, so I'm not sure it can be done in the same way as the black one was done. Maybe it can be solved in some other way though.
2. It seems from the docs that it is only able to format files and cannot (?) receive a string/input stream

pelme replied:
Feel free to drop the formatting altogether now! It is better to merge this as is and then we can add more features later.

Using a subprocess may be more straightforward to support both. It is enough just to have black or ruff on PATH to make it work:

```pycon
Python 3.12.3 (main, Apr  9 2024, 08:09:14) [Clang 16.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> subprocess.run(['black', '-q', '-'], input='def foo(a   = 1):pass'.encode('utf8')).stdout
def foo(a=1):
    pass
>>> subprocess.run(['ruff', 'format', '-'], input='def foo(a   = 1):pass'.encode('utf8')).stdout
def foo(a=1):
    pass
>>>
```

OleJoik replied:
I added this in the latest commit.

OleJoik commented on htpy/cli/cli.py:
```diff
@@ -0,0 +1,79 @@
+    print(f"  shorthand id class: {shorthand_id_class}")
+    print("\n>>>>>>>>>>>>>>>>>>")
+    print(">>> paste html >>>")
+    print(">>>>>>>>>>>>>>>>>>\n")
```
Alright, now it takes file/stdin input instead of my hacky terminal paste thing.

I decided to drop the verbose mode and just simplify the out stream conciderably instead, as I don't really feel that printing the selected options are necessary?

OleJoik commented:
Alright, I think I'm happy with the current state at this point. I've taken the liberty to resolve some of the comment threads (don't know if that is bad practice of me to do..?), maybe you'll do a final sweep and clarify if there is anything else you want changed?

Ill add a simple docs page

OleJoik commented:
One thing I thought about might be an option to add imports for used elements in the output. Might be nice to save even a couple of more seconds.

That should be quick to add if you want it, but might also be another PR perhaps?

pelme commented on htpy/html2htpy.py:
```diff
@@ -0,0 +1,284 @@
+
+    args = parser.parse_args()
+
+    if args.input == sys.stdin and select.select([sys.stdin], [], [], 0.1)[0]:
```
what does the select call do here? would it work without it?

OleJoik replied:
I must admit I'm fairly light on this topic. The internet is my guide.

My understanding, however, is that it checks if there is input on the stdin stream (or if input shows up within a deadline of 0.1 seconds). On my system, without it, the command hangs if you run it without piping input (if you did not pass a file). With it, python detects if there is no input piped AND no file provided, so it can enter the else block, error out and exit.

see this SO-answer: [https://stackoverflow.com/a/3763257](https://stackoverflow.com/a/3763257)

According to SO answer this might only work on unix systems, and I've only tested it on linux.

If you have any advice I am open ears. Otherwise I can investigate this topic further tomorrow.

OleJoik replied:
Maybe the command hanging without input is actually correct behaviour. I see that grep behaves the same way, as also referenced in the SO post.

pelme replied:
Hanging is how most commands work and what is generally expected. It also allows for pasting things directly into a terminal (I use this trick often). I agree it may seem confusing and weird that it just hangs but that is the convention.

For instance, running `black -` and then typing `a=1<enter><ctrl+d>` looks like this:
```
$ black -
a=1
a = 1
reformatted -

All done! ✨ 🍰 ✨
1 file reformatted.
```

OleJoik replied:
Excellent, learning new things here! I removed the select call and it works like you'd expect!

OleJoik commented:
Also added some docs for review!

OleJoik commented:
I updated the formatting to use subprocess as per your suggestion. Seems to work great. Also removed the `extras` dependency group, as I suppose the users mostly can install formatters themselves?

pelme reviewed (changes requested):
Thanks a lot for putting in the effort to finalize this. The quality of docs and tests are really nice. This feels like a very nice addition and it works really smooth!

OleJoik commented on htpy/html2htpy.py:
```diff
@@ -160,26 +191,21 @@ def serialize_python(self, shorthand_id_class: bool = False, format: bool = Fals
-                o, mode=black.FileMode(line_length=80, magic_trailing_comma=False)
-            )
+        if formatter:
+            return formatter.format(o)
```
I actually implemented it like that first, and switched to this, haha. I'll switch back!

OleJoik replied:
At second thought, I will actually skip this one for now. Changing it is actually not completely straight forward - would have to either pass a formatter instance to every python call of html2htpy function call (e.g. in every test) or deal with some default shenanigans. I made the default a lambda returning a Formatter, but it all got a bit out of hand so I discarded that commit..

OleJoik commented on docs/html2htpy.md:
```diff
@@ -0,0 +1,188 @@
+
+# Convert HTML to HTPY
```
Yes, I'll do!

OleJoik commented on htpy/html2htpy.py:
```diff
@@ -0,0 +1,366 @@
+
+
+if __name__ == "__main__":
+    main()
```
Thats ok, Ill remove it

OleJoik commented on tests/test_html2htpy.py:
```diff
@@ -0,0 +1,311 @@
+        </div>
+    """
+
+    actual = html2htpy(input, shorthand_id_class=True, formatter=BlackFormatter())
```
Agreed, I will make some changes here, and make sure that all the formatters are covered.

OleJoik replied:
I removed the formatting from most of the tests, but kept it for the final 2 "larger tests" to make the expected value more readable in the test. Additionally added a couple of tests with the `RuffFormatter`.

OleJoik commented on docs/html2htpy.md:
```diff
@@ -7,15 +7,16 @@ html into python code (htpy!).
-  -s, --shorthand  Use shorthand syntax for class and id attributes
-  -f, --format     Format output code (requires black installed)
+  -h, --help            show this help message and exit
+  -s, --shorthand       Use shorthand syntax for class and id attributes
```
I guess the point is that you prefer the default to be the shorthand syntax, which is totally legit. But I don't know if I like the negating "no-" for a boolean value, I find it confusing. Do you think `-e` / `--explicit` with a good help text is acceptable?

OleJoik replied:
I changed it to `--no-shorthand`!

OleJoik commented:
Alright, I think I've addressed the comments.

Thank you for all the thorough feedback and timely followup by the way.

OleJoik commented:
There was this little thing I wanted to add:

```
## Detect htpy imports

If you pass the `-i`/`--imports` flag, htpy elements detected will be included as 
imports in the output. For example:

```py title="$ html2htpy --imports example.html"
from htpy import p, section

section("#main-section.hero.is-link")[
    p(".subtitle.is-3.is-spaced")["Welcome"]
]
```

In the end, I couldn't help myself, so I added it now. Ill stop adding new things now, promise.

pelme commented:
> There was this little thing I wanted to add:
> ## Detect htpy imports
> 
> If you pass the `-i`/`--imports` flag, htpy elements detected will be included as imports in the output. For example:
> 
> `$ html2htpy --imports example.html`
> 
> ```python
> from htpy import p, section
> 
> section("#main-section.hero.is-link")[
>     p(".subtitle.is-3.is-spaced")["Welcome"]
> ]
> ```
> 
> In the end, I couldn't help myself, so I added it now. Ill stop adding new things now, promise.

I like this a lot!! I also think that the imports should be added by default. Most people will probably `pbpaste | html2htpy` anyways and then it is easy to just not copy the import if it is not needed anyways.

Lately I have been using `import htpy as h` and find it nice. No need to fiddle with imports all the time. 

What do you think about `--imports=yes|h|no` with yes being the default: `from htpy import div, img, ...`. And `--imports=h`would be `import htpy as h` / `h.div[h.span["Hi"]]`. `--imports=no` would be no imports at all.

We can add `--imports=h` later after this is merged but that would allow us to add it going forward.

OleJoik commented:
I added it!

## Import options

You have a couple of options regarding imports with the `-i`/`--imports` flag. 
Options are `yes` (default), `h`, `no`. 

#### Module import of htpy: `--imports=h`

Some people prefer to `import htpy as h` instead of importing individual elements from htpy.
If this is you, you can use the `--imports=h` option to get corresponding output when using `html2htpy`.

```
$ html2htpy --imports=h example.html
```

```py
import htpy as h

h.section("#main-section.hero.is-link")[
    h.p(".subtitle.is-3.is-spaced")["Welcome"]
]
```

OleJoik commented:
fyi, in addition to the unit tests, I did some manual testing with a bunch of large jinja templates that I had laying around, and it seems to behave well even for fairly ugly templates with lots of weird jinja shit.

OleJoik commented:
I also did some proof-reading of the docs. Think I'm pretty much good from my side now.

pelme commented:
Thanks A LOT for this and all the good discussions! I opened a PR with some small changes, please have a look #27 !
